### ISSUE TITLE
Fieldmaps causing fMRIprep to get crash

### ISSUE THREAD

**Issue by NickESouter** [nipreps/fmriprep#2947]
### What happened?

I'm currently working with unprocessed 3T MRI and fMRI data from the human connectome project. I'm using:

- T1w scan, with corresponding 'phasediff', 'magnitude1', and 'magnitude2' fieldmaps
- T2w scan, with corresponding 'phasediff', 'magnitude1', and 'magnitude2' fieldmaps
- Four functional task runs, each with corresponding RL and LR epi spin echo fieldmaps (two opposite phase encoding directions for each run). Each also has a corresponding SBRef scan.

The dataset comes up as totally fine on the BIDS validator, and when I run fMRIPrep with the fieldmaps removed from the directory, everything works perfectly. If I try again with the fieldmaps returned to their 'fmap' subdirectory, the job appears to crash before fMRIPrep even starts running. By looking at my other log files, the step it's getting stuck on normally takes 20 seconds on so, and the next step should be acknowledging any fieldmaps/SBRef files available for each functional run.

Between successful runs and my unsuccessful ones with fieldmaps, nothing in the script has changed, just the presence of fmap files in the BIDS directory. Clearly something in the fieldmaps is causing things to stick here but I'm not sure what it is. Does fMRIPrep expect some specific BIDS-compatible naming convention or sidecar information that I might be missing? I've provided all the available metadata I can find, as shown here:

Example NIFTI filename: 'sub-100307_acq-EmotionLR_dir-LR_epi.nii.gz'

Example JSON metadata: {"IntendedFor": "bids::sub-100307/func/sub-100307_task-Emotion_dir-LR_bold.nii.gz", "Manufacturer": "Siemens", "ManufacturersModelName": "Connectome Skyra", "MagneticFieldStrength": 3, "ReceiveCoilName": "Standard 32-Channel Siemens Receive Head Coil", "ReceiveCoilActiveElements": "HEA;HEP", "PulseSequenceType": "Spin Echo Field Map", "NonlinearGradientCorrection": "false", "MRAcquisitionType": "2D", "MTState": "false", "EffectiveEchoSpacing": 0.00058, "EchoTime": 0.058, "FlipAngle": 90, "MultibandAccelerationFactor": 1, "TaskName": "Emotion", "RepetitionTime": 7.06, "TotalReadoutTime": 0.08346, "B0FieldIdentifier": "EmotionLR_bold_fmap", "PhaseEncodingDirection": "i"}

Or is it possible this might be a memory issue, not related to fMRIPrep itself?

Any advice would be really appreciated!

Nick

### What command did you use?

```shell
singularity run --cleanenv \
    -B ${DATA_DIR}:/data \
    -B ${OUT_DIR}/:/out \
    -B ${SCRATCH_DIR}:/wd \
    -B ${LICENSE}:/license \
    /research/cisc2/shared/fmriprep_singularity/fmriprep_22.1.1.simg \
    --skip_bids_validation \
    --participant-label ${SUBJECT} \
    --omp-nthreads 1 --nthreads 5 --mem_mb 30000 \
    --output-spaces MNI152NLin6Asym:res-2 MNI152NLin2009cAsym fsaverage T1w fsLR:den-32k \
    --fs-license-file /license \
    --work-dir /wd \
    --cifti-output 91k \
    --ignore slicetiming \
    --track-carbon \
    --country-code GBR \
    --random-seed 1234 \
    --skull-strip-fixed-seed \
    --clean-workdir \
    /data /out/ participant
```

### What version of fMRIPrep are you running?

22.1.1.

### How are you running fMRIPrep?

Singularity

### Is your data BIDS valid?

Yes

### Are you reusing any previously computed results?

No

### Please copy and paste any relevant log output.

```shell
Number of subjects = 1
This is the task id 1
this is i 0
sub-100307
230210-14:37:04,973 nipype.workflow IMPORTANT:
	 CodeCarbon tracker started ...
230210-14:37:04,974 nipype.workflow IMPORTANT:
	 Using country_iso_code: GBR
230210-14:37:04,974 nipype.workflow IMPORTANT:
	 Saving logs at: /out/logs
230210-14:37:18,647 nipype.workflow IMPORTANT:
	 Running fMRIPrep version 22.1.1

         License NOTICE ##################################################
         fMRIPrep 22.1.1
         Copyright 2022 The NiPreps Developers.
         
         This product includes software developed by
         the NiPreps Community (https://nipreps.org/).
         
         Portions of this software were developed at the Department of
         Psychology at Stanford University, Stanford, CA, US.
         
         This software redistributes the versioneer Python package, which is
         Public domain source code.
         
         This software is also distributed as a Docker container image.
         The bootstraping file for the image ("Dockerfile") is licensed
         under the MIT License.
         
         This software may be distributed through an add-on package called
         "Docker Wrapper" that is under the BSD 3-clause License.
         #################################################################
230210-14:37:18,723 nipype.workflow IMPORTANT:
	 Building fMRIPrep's workflow:
           * BIDS dataset path: /data.
           * Participant list: ['100307'].
           * Run identifier: 20230210-143658_66b8c928-cdde-44a9-a44b-3822885560c3.
           * Output spaces: MNI152NLin6Asym:res-2 MNI152NLin2009cAsym:res-native fsaverage:den-164k T1w fsLR:den-32k.
           * Pre-run FreeSurfer's SUBJECTS_DIR: /out/sourcedata/freesurfer.
Done
```

### Additional information / screenshots

My BIDS filetree:

|   dataset_description.json
|   participants.json
|   participants.tsv
|   README.txt
|   
\---sub-100307
    +---anat
    |       sub-100307_T1w.json
    |       sub-100307_T1w.nii.gz
    |       sub-100307_T2w.json
    |       sub-100307_T2w.nii.gz
    |       
    +---fmap
    |       sub-100307_acq-EmotionLR_dir-LR_epi.json
    |       sub-100307_acq-EmotionLR_dir-LR_epi.nii.gz
    |       sub-100307_acq-EmotionLR_dir-RL_epi.json
    |       sub-100307_acq-EmotionLR_dir-RL_epi.nii.gz
    |       sub-100307_acq-EmotionRL_dir-LR_epi.json
    |       sub-100307_acq-EmotionRL_dir-LR_epi.nii.gz
    |       sub-100307_acq-EmotionRL_dir-RL_epi.json
    |       sub-100307_acq-EmotionRL_dir-RL_epi.nii.gz
    |       sub-100307_acq-MotorLR_dir-LR_epi.json
    |       sub-100307_acq-MotorLR_dir-LR_epi.nii.gz
    |       sub-100307_acq-MotorLR_dir-RL_epi.json
    |       sub-100307_acq-MotorLR_dir-RL_epi.nii.gz
    |       sub-100307_acq-MotorRL_dir-LR_epi.json
    |       sub-100307_acq-MotorRL_dir-LR_epi.nii.gz
    |       sub-100307_acq-MotorRL_dir-RL_epi.json
    |       sub-100307_acq-MotorRL_dir-RL_epi.nii.gz
    |       sub-100307_acq-T1w_magnitude1.nii.gz
    |       sub-100307_acq-T1w_magnitude2.nii.gz
    |       sub-100307_acq-T1w_phasediff.json
    |       sub-100307_acq-T1w_phasediff.nii.gz
    |       sub-100307_acq-T2w_magnitude1.nii.gz
    |       sub-100307_acq-T2w_magnitude2.nii.gz
    |       sub-100307_acq-T2w_phasediff.json
    |       sub-100307_acq-T2w_phasediff.nii.gz
    |       
    \---func
            sub-100307_task-Emotion_dir-LR_bold.json
            sub-100307_task-Emotion_dir-LR_bold.nii.gz
            sub-100307_task-Emotion_dir-LR_events.tsv
            sub-100307_task-Emotion_dir-LR_sbref.json
            sub-100307_task-Emotion_dir-LR_sbref.nii.gz
            sub-100307_task-Emotion_dir-RL_bold.json
            sub-100307_task-Emotion_dir-RL_bold.nii.gz
            sub-100307_task-Emotion_dir-RL_events.tsv
            sub-100307_task-Emotion_dir-RL_sbref.json
            sub-100307_task-Emotion_dir-RL_sbref.nii.gz
            sub-100307_task-Motor_dir-LR_bold.json
            sub-100307_task-Motor_dir-LR_bold.nii.gz
            sub-100307_task-Motor_dir-LR_events.tsv
            sub-100307_task-Motor_dir-LR_sbref.json
            sub-100307_task-Motor_dir-LR_sbref.nii.gz
            sub-100307_task-Motor_dir-RL_bold.json
            sub-100307_task-Motor_dir-RL_bold.nii.gz
            sub-100307_task-Motor_dir-RL_events.tsv
            sub-100307_task-Motor_dir-RL_sbref.json
            sub-100307_task-Motor_dir-RL_sbref.nii.gz

_No response_

**Comment by NickESouter** [comment]
Hi Chris,

Thanks for taking a look at this! Apologies, I should have edited this post to say that I have recently tried implementing both of these changes, and unfortunately my script is still crashing in the same place. Here's the most recent version of my fieldmap JSON sidecars:

{"IntendedFor": "func/sub-100307_task-Emotion_dir-LR_bold.nii.gz", "Manufacturer": "Siemens", "ManufacturersModelName": "Connectome Skyra", "MagneticFieldStrength": 3, "ReceiveCoilName": "Standard 32-Channel Siemens Receive Head Coil", "ReceiveCoilActiveElements": "HEA;HEP", "PulseSequenceType": "Spin Echo Field Map", "NonlinearGradientCorrection": "false", "MRAcquisitionType": "2D", "MTState": "false", "EffectiveEchoSpacing": 0.00058, "EchoTime": 0.058, "FlipAngle": 90, "MultibandAccelerationFactor": 1, "TaskName": "Emotion", "RepetitionTime": 7.06, "TotalReadoutTime": 0.08346, "B0FieldIdentifier": "EmotionLR_bold_fmap", "PhaseEncodingDirection": "i"}

And for the respective functional run:

{"Manufacturer": "Siemens", "ManufacturersModelName": "Connectome Skyra", "MagneticFieldStrength": 3, "ReceiveCoilName": "Standard 32-Channel Siemens Receive Head Coil", "ReceiveCoilActiveElements": "HEA;HEP", "PulseSequenceType": "Gradient Echo EPI", "NonlinearGradientCorrection": "false", "MRAcquisitionType": "2D", "MTState": "false", "EffectiveEchoSpacing": 0.00058, "EchoTime": 0.0331, "SliceTiming": [0.0, 0.39, 0.0775, 0.4675, 0.1575, 0.545, 0.235, 0.6225, 0.3125, 0.0, 0.39, 0.0775, 0.4675, 0.1575, 0.545, 0.235, 0.6225, 0.3125, 0.0, 0.39, 0.0775, 0.4675, 0.1575, 0.545, 0.235, 0.6225, 0.3125, 0.0, 0.39, 0.0775, 0.4675, 0.1575, 0.545, 0.235, 0.6225, 0.3125, 0.0, 0.39, 0.0775, 0.4675, 0.1575, 0.545, 0.235, 0.6225, 0.3125, 0.0, 0.39, 0.0775, 0.4675, 0.1575, 0.545, 0.235, 0.6225, 0.3125, 0.0, 0.39, 0.0775, 0.4675, 0.1575, 0.545, 0.235, 0.6225, 0.3125, 0.0, 0.39, 0.0775, 0.4675, 0.1575, 0.545, 0.235, 0.6225, 0.3125], "FlipAngle": 52, "MultibandAccelerationFactor": 8, "TaskName": "Emotion", "RepetitionTime": 0.72, "B0FieldSource": "EmotionLR_bold_fmap", "PhaseEncodingDirection": "i"}

Do these look like they're setup correctly now? If so, it may be that I'm experiencing some larger issue with the files themselves or with our HPC.

Nick

**Comment by effigies** [comment]
Can you try running `sdcflows-find-estimators` on your dataset? It should be inside your Singularity container. Something like:

```
singularity exec ... /research/cisc2/shared/fmriprep_singularity/fmriprep_22.1.1.simg sdcflows-find-estimators /data -s 100307
```

If that fails, then would you be willing to share your (partial) dataset? It's likely a quick enough bug to find with concrete data to work on. You could truncate the NIfTIs to just the headers.

**Comment by NickESouter** [comment]
I've tried running this sdcflows command but can't seem to get that working, strangely I'm being told that the 'singularity' command is not found... I'm sure I'm just entering this command in the wrong place, I'll look into this and chat with people in my lab.

I'm very happy to share the data for this one subject, given that it's open access through HCP anyway. I could share the full data although the zip folder seems too large to upload here, would there be an easier place to share it with you? If not, I can truncate to the headers.

Thanks so much,

Nick

**Comment by effigies** [comment]
I feel like HCP has been BIDS-ified a couple times. Did you start with something that could be fetched publicly, such that we just need to update the metadata to match yours? I haven't played with the data myself in probably 8-9 years, so I don't know what the current best practices are.

**Comment by NickESouter** [comment]
I think one of the main issues that's getting in the way of troubleshooting is that very few people seem to have worked with the unprocessed version of HCP data before, at least as far as I've found. Rather, people tend to work with the data that's already been through the HCP pipeline. I've found cases of scripts available to convert preprocessed HCP data to BIDS (e.g., https://github.com/datalad-datasets/hcp-structural-preprocessed), but no cases of converted unprocessed data.

There's also this case of a tool designed for this very purpose (hcp2bids: https://github.com/suyashdb/hcp2bids), but unfortunately this doesn't quite get the data BIDS-compatible, and even after some tweaking appears to encounter the same issue I've been facing.

So I essentially started with the raw unprocessed data from the HCP young adult dataset, then wrote a Python script to convert the relevant files to BIDS. Because this unprocessed data comes in the form of NIFTIs rather than DICOMs, I had to extract metadata information manually from the HCP reference manual, I just took everything I could find to correspond to as many of the BIDS categories as possible (these are all listed in my examples JSONs above). HCP has a policy of not sharing metadata for the original DICOMs. If you know of anyone who has successfully run HCP YA data through fMRIPrep, it would be a massive help to be able to see some examples of the data. As you say, I think this would basically put the issue to rest!

**Comment by effigies** [comment]
Well, if you can upload a zip file with truncated headers (even just one run of one task) and all the JSON/TSV files, that would be really helpful. You can privately share with me using this user name @ gmail.

**Comment by dmoracze** [comment]
Our team has "successfully" converted HCP to BIDS and run fmriprep. By "successfully" I mean the pipeline runs without error. However, we've had a very difficult time trying to integrate the fmaps. The fmriprep derivatives of the functional data are unusable, with massive warping issues. We've used fmriprep version 20.2.0 and 21.0.2, the latter is a bit better but still has enough warping issues to be unusable. 

To convert to BIDS, we've used hcp2bids, as @NickESouter mentioned. However, we've been trying to follow a pipeline similar to [this paper](https://arxiv.org/abs/2205.15581). The authors used [this script](https://github.com/athms/interpreting-brain-decoding-models/blob/master/scripts/rearrange_hcp-source_to_BIDS.py) to convert their data, so we ended up following that. However, it does not copy the fmap images.

**Comment by hough129** [comment]
Hi @NickESouter and @dmoracze - I work for Damien Fair at the Masonic Institute of the Brain at UMN. We were about to convert HCP-YA to BIDS, and @effigies pointed out your GitHub issue. Would you two be willing to meet with @ericfeczko and me to discuss possibly joining forces to figure out these last issues with the BIDS and processing?

**Comment by NickESouter** [comment]
Thanks @effigies, I'll try to get the truncated form of the data over to you by this afternoon!

@dmoracze Thank you Dustin, it's useful to know that there are wider issues with the fieldmaps here and not just a local issue. For various reasons we have been considering using an alternative dataset, so it's good to know that the fieldmaps don't run perfectly even when the pipeline is able to reach and end. I'll make a note of this paper and script though, I hadn't come across them before!

Hi @hough129, definitely happy to have a meeting to chat through this, and I can share my conversion script with you if helpful (obviously it's not actually producing data that can run through fMRIPrep, but it is fully BIDS compatible as far as the BIDS validator is concerned...). Feel free to reach out directly and we can set up a time to meet, you can find my contact info here; https://profiles.sussex.ac.uk/p596400-nick-souter

**Comment by dmoracze** [comment]
@NickESouter @ericfeczko @hough129 

Open to discussing more so we aren't doubling our efforts. @ericfeczko - I'm familiar with the abcd-hcp-pipeline, it's a great tool. Though it would be great if users could also have the option use fmriprep for their HCP data.

**Comment by effigies** [comment]
Our bottleneck right now is developer hours for people comfortable with both nipype and SDC. There are some orientation handling issues that are probably at the heart of the failure with RL PEPolar and some sagittal acquisitions I recently got my hands on, so I'm hopeful that we can get to it soon. But if there's anybody else in our respective groups who want to take on debugging some specific use cases, that will make it go much faster.

**Comment by ericfeczko** [comment]
@dmoracze we feel the same way, derived data should be accessible across multiple formats to facilitate accessibility and future studies on processing harmonization. In fact our team has been working with @effigies on feature enhacements, etc. with fmriprep for that reason.

@effigies is right, if we can pool efforts, we might be able to devote more cycles to resolving the underlying issue and getting fmriprep moving.

**Comment by effigies** [comment]
Hi all, just to let you know I've been looking at this and I have a proposal here: https://github.com/nipreps/sdcflows/pull/339

I updated @NickESouter's metadata to use PhaseEncodingDirection of `i`/`i-` (both were `i`) and got the following fieldmap:

![image](https://user-images.githubusercontent.com/83442/224346684-25fcfeae-6059-488e-8d57-b628863be794.png)

Which leads to the following before/after:

![Screenshot from 2023-03-10 09-51-09](https://user-images.githubusercontent.com/83442/224347017-1527a5ad-27df-4221-be35-3c81c38e0a56.png)
![Screenshot from 2023-03-10 09-51-00](https://user-images.githubusercontent.com/83442/224347022-5a6659d1-4ece-4082-ba80-4700b5151094.png)

It's not perfect correction, but it is at least in the right direction. I'm not sure what your experience has been with TOPUP directly?

**Comment by NickESouter** [comment]
Wow that was a very sloppy error on my part with the metadata... I can see now that an error in my conversion script was in some instances setting phase encoding direction according to the task itself, rather than the fieldmap. Thanks so much for catching this @effigies. I've now been able to get fMRIprep running on this data with the updated files.

To me eye the 'after' image there doesn't look very different to the 'before', although I have to say I don't have much basis for comparison myself!

**Comment by ericfeczko** [comment]
@NickESouter no worries, datasets with large scale are difficult to version control, its why we think FAIR principles are so important. Thanks for catching this, we will be sure to adjust it when running on our side :)

@effigies Yeah, this is where I'm interested to see how the HCP specific pipelines handle this data. There may be facets that could be incorported into fmriprep to handle the unusual L/R distortion. Though scaling it so that its intuitive may also require a small BEP to incorporate other data/metadata. 

We have a few people on this, one of them happens to be on break this week, so we might be able to compare some outputs late next week. Given how this study was designed, if the data can be processed both well and reproducibly it would represent a happy medium between the two paths towards neuroscience -- getting this right is kind of a holy grail for available datasets ;) 

I for one would love to share such a grail with the community 😁

**Comment by NickESouter** [comment]
Hi all,

Sorry to open up this very old thread again! Just a note to say that after getting my HCP young adult to BIDS conversion script working (Thanks to you all for your help and advice on this!), I've cleaned it up, and updated such that it'll now convert 3T and 7T MRI data, as well as diffusion data to BIDS. I've successfully run a sample of the 3T data through fMRIPrep.

I've made it available here, just in case it's helpful: https://github.com/NickESouter/HCPinBIDS

Happy to take another look at it if there's any issues :)
